sklearn/linear_model/_ridge.py

```diff
@@ -2082,11 +2094,17 @@ def fit(self, X, y, sample_weight=None, score_params=None):
         -------
         self : object
         """
+        xp, is_array_api = get_namespace(X, y, sample_weight)
+        device_kwargs = {"device": device(X)} if is_array_api else {}
+        if sparse.issparse(X):
+            dtype = np.float64
+        else:
+            dtype = [xp.float64, xp.float32]
```

Comment:
Contrary to what I said in this morning meeting, I think we might want to implement the following logic:

- if the input namespace/device supports `xp.float64` upcasting, then do the upcast (as we currently do with NumPy)
- if not (e.g. pytorch + MPS device combination), accept that we have degraded numerical performance, adjust the tolerance in the tests accordingly and document this limited numerical precision guarantee in our Array API doc.


I think this is the strategy we are leaning towards in the review of #27113. During the review of the `r2_score` PR, I believe that @adrinjalali preferred that approach.

In a future PR, we might decide to drop the float32 -> float64 upcast in general for this estimator (as it's silently trigger a potentially very large and unexpected memory allocation which is a usability problem in itself) but I would rather make this decision independently of Array API support.

Reply:
how would you recommend I check if the upcasting is possible? should I temporarily copy the `max_precision_float_dtype` and `supported_float_dtypes` changes from [27113](https://github.com/scikit-learn/scikit-learn/pull/27113/files#diff-6264adab84df6fafdb2d950141783b52b718da1e1be773daae7b5921b2556f94R338) until it is merged? or is there already a utility in scikit-learn for checking that which I missed?

Reply:
Feel free to copy with a TODO comment to remove redundant code once #27113 is merged to be able to decouple the 2 reviews.

Reply:
when we do the upcast with what precision should we store the coefficients and intercept? I guess for prediction we do not need the extra precision so we should use X's original dtype?